libs/geo/src/main/java/org/elasticsearch/geometry/utils/SpatialEnvelopeVisitor.java

```diff
@@ -0,0 +1,302 @@
+ *         and the smaller bounding box is chosen. It is possible to disable the wrapping of the longitude.
+ * </ul>
+ * Usage of this is as simple as:
+ * <code>
+ *     Optional&lt;Rectangle&gt; bbox = SpatialEnvelopeVisitor.visit(geometry);
+ *     if (bbox.isPresent()) {
+ *         Rectangle envelope = bbox.get();
+ *         // Do stuff with the envelope
+ *     }
+ * </code>
+ * It is also possible to create the inner <code>PointVisitor</code> separately, as well as use the visitor for multiple geometries.
+ * <code>
+ *     PointVisitor pointVisitor = new CartesianPointVisitor();
+ *     SpatialEnvelopeVisitor visitor = new SpatialEnvelopeVisitor(pointVisitor);
+ *     for (Geometry geometry : geometries) {
+ *         geometry.visit(visitor);
+ *     }
+ *     if (visitor.isValid()) {
+ *         Rectangle envelope = visitor.getResult();
+ *         // Do stuff with the envelope
+ *     }
+ * </code>
+ * Code that wishes to modify the behaviour of the visitor can implement the <code>PointVisitor</code> interface,
+ * or extend the existing implementations.
+ */
+public class SpatialEnvelopeVisitor implements GeometryVisitor<Boolean, RuntimeException> {
+
+    private final PointVisitor pointVisitor;
+
+    public SpatialEnvelopeVisitor(PointVisitor pointVisitor) {
+        this.pointVisitor = pointVisitor;
+    }
+
+    /**
+     * Determine the BBOX without considering the CRS or wrapping of the longitude.
+     * Note that incoming BBOX's that do cross the dateline (minx>maxx) will be treated as invalid.
+     */
+    public static Optional<Rectangle> visit(Geometry geometry) {
+        var visitor = new SpatialEnvelopeVisitor(new CartesianPointVisitor());
+        if (geometry.visit(visitor)) {
+            return Optional.of(visitor.getResult());
+        }
+        return Optional.empty();
+    }
+
+    /**
+     * Determine the BBOX assuming the CRS is geographic (eg WGS84) and optionally wrapping the longitude around the dateline.
+     */
+    public static Optional<Rectangle> visit(Geometry geometry, boolean wrapLongitude) {
+        var visitor = new SpatialEnvelopeVisitor(new GeoPointVisitor(wrapLongitude));
+        if (geometry.visit(visitor)) {
+            return Optional.of(visitor.getResult());
+        }
+        return Optional.empty();
+    }
+
+    public Rectangle getResult() {
+        return pointVisitor.getResult();
+    }
+
+    /**
+     * Visitor for visiting points and rectangles. This is where the actual envelope calculation happens.
+     * There are two implementations, one for cartesian coordinates and one for geographic coordinates.
+     * The latter can optionally wrap the longitude around the dateline.
+     */
+    public interface PointVisitor {
+        void visitPoint(double x, double y);
+
+        void visitRectangle(double minX, double maxX, double maxY, double minY);
+
+        boolean isValid();
+
+        Rectangle getResult();
```

Comment:
Since this can return `null`, it should return `Optional<Rectangle>` (Or at the very least be marked with `@Nullable`, but `Optional` is better)

Reply:
It should never return null. I removed the one case where it does.